File: dev_zero_client.c
///////////////////////////////////////////////////////////////////////////
//This is a simple client which is intended to connect to dev_null_server
//once connected this client will push 4G of data read from /dev/zero
//
//       Usage: ./dev_zero_client <server_ip> <port>
//
///////////////////////////////////////////////////////////////////////////

#include <stdio.h>
#include <arpa/inet.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <stddef.h>
#include <fcntl.h>
#include <stdint.h>

#define BUFFER_SIZE (1 << 20) // 1 MB buffer size

int main(int argc, char *argv[]) {
    int sockfd, portno, n;
    struct sockaddr_in serv_addr;
    char buffer[BUFFER_SIZE];
    int zero_fd;
    int total_bytes_sent = 0;
    unsigned long bytes_to_send = 4u * 1014u * 1024u * 1024u; // 4 GB

    // Open /dev/zero for reading
    if ((zero_fd = open("/dev/zero", O_RDONLY)) < 0) {
        perror("Error opening /dev/zero");
        exit(1);
    }

    // Create a socket
    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        perror("Error creating socket");
        exit(1);
    }

    // Set up the server address
    bzero((char *) &serv_addr, sizeof(serv_addr));
    portno = atoi(argv[2]);
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(argv[1]);
    serv_addr.sin_port = htons(portno);

    // Connect to the server
    if (connect(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
        perror("Error connecting to server");
        exit(1);
    }

    // Read from /dev/zero and write to the socket
    while (total_bytes_sent < bytes_to_send) {
        int bytes_read = read(zero_fd, buffer, sizeof(buffer));
        if (bytes_read < 0) {
            perror("Error reading from /dev/zero");
            exit(1);
        }
        int bytes_sent = write(sockfd, buffer, bytes_read);
        if (bytes_sent < 0) {
            perror("Error writing to socket");
            exit(1);
        }
        total_bytes_sent += bytes_sent;
    }

    // Close the socket and /dev/zero file descriptor
    close(sockfd);
    close(zero_fd);

    return 0;
}

File: dev_zero_client.py
///////////////////////////////////////////////////////////////////////////
//This is a simple client which is intended to connect to dev_null_server
//once connected this client will push 4G of data read from /dev/zero
//
//       python dev_zero_client.py <server_ip> <port>
//
// Author: Michael Colombo
// Date: 5/7/2023 
//
///////////////////////////////////////////////////////////////////////////

#! /usr/bin/env python

import socket
import sys

HOST = sys.argv[1]
PORT = int (sys.argv[2])

# Create a TCP/IP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Connect the socket to a remote server
sock.connect((HOST, PORT))
print("Connected to {}:{}".format(HOST, PORT))

# Open /dev/zero for reading
with open('/dev/zero', 'rb') as zero_file:
    # Read 4GB of data from /dev/zero
    bytes_to_read = 4 * 1024 * 1024 * 1024
    while bytes_to_read > 0:
        # Read data from /dev/zero in chunks of 1MB
        chunk_size = min(bytes_to_read, 1048576)
        data = zero_file.read(chunk_size)

        # Write the data to the socket
        sock.sendall(data)

        # Decrement the bytes left to read
        bytes_to_read -= chunk_size

# Close the socket
print("Closing connection to {}:{}".format(HOST, PORT))
sock.close()

